### ISSUE TITLE
Unable to retrieve window handles after connecting to existing Firefox instance

### ISSUE THREAD

**Issue by Danp2** [mozilla/geckodriver#1790]
## System

* Version: 0.27.0
* Platform: Windows 10 64bit
* Firefox: 81.0.1
* Selenium: None

## Testcase

- Launch geckodriver with the parameters `--log trace --connect-existing  --marionette-port 2828`
- Attempt to retrieve handles for existing tabs

## Stacktrace

N/A

## Trace-level log

```
1602454773202   geckodriver::marionette DEBUG   Connection to Marionette established on 127.0.0.1:2828.
1602454773210   webdriver::server       DEBUG   <- 200 OK {"value":{"sessionId":"d1495a10-a17c-4ef9-b466-dab022820eea","capabilities":{"acceptInsecureCerts":false,"browserName":"firefox","browserVersion":"81.0.1","moz:accessibilityChecks":false,"moz:buildID":"20200930150533","moz:geckodriverVersion":"0.27.0","moz:headless":false,"moz:processID":63620,"moz:profile":"C:\\Users\\danpollak\\AppData\\Roaming\\Mozilla\\Firefox\\Profiles\\9e7wtvmu.test","moz:shutdownTimeout":60000,"moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"windows","platformVersion":"10.0","rotatable":false,"setWindowRect":true,"strictFileInteractability":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"unhandledPromptBehavior":"dismiss and notify"}}}
1602454773231   webdriver::server       DEBUG   -> GET /session/d1495a10-a17c-4ef9-b466-dab022820eea/url
1602454773233   webdriver::server       DEBUG   <- 200 OK {"value":"https://github.com/mozilla/geckodriver/issues"}
1602454773240   webdriver::server       DEBUG   -> POST /session/d1495a10-a17c-4ef9-b466-dab022820eea/url {"url":"https://github.com/mozilla/geckodriver/issues"}
1602454774023   webdriver::server       DEBUG   <- 200 OK {"value":null}
1602454774034   webdriver::server       DEBUG   -> GET /session/d1495a10-a17c-4ef9-b466-dab022820eea/window/handles
1602454774060   webdriver::server       DEBUG   <- 500 Internal Server Error {"value":{"error":"unknown error","message":"TypeError: browser.browsingContext is null","stacktrace":"GeckoDriver.prototype.getIdForBrowser@chrome://marionette/content/driver.js:1477:15\nget@chrome://marionette/content/driver.js:273:28\nGeckoDriver.prototype.getWindowHandles@chrome://marionette/content/driver.js:1517:3\ndespatch@chrome://marionette/content/server.js:305:40\nexecute@chrome://marionette/content/server.js:275:16\nonPacket/<@chrome://marionette/content/server.js:248:20\nonPacket@chrome://marionette/content/server.js:249:9\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:501:20\n"}}
```

**Comment by Danp2** [comment]
This worked as expected in FF 80.x and earlier releases

**Comment by whimboo** [comment]
Could you please attach a [trace-level log] from geckodriver? Read more about reporting actionable bugs in our [contribution guidelines].

[trace-level log]: https://firefox-source-docs.mozilla.org/testing/geckodriver/TraceLogs.html
[contribution guidelines]: https://firefox-source-docs.mozilla.org/testing/geckodriver/Bugs.html

**Comment by Danp2** [comment]
```
geckodriver.exe -vv --connect-existing  --marionette-port 2828
1602513389326   geckodriver     INFO    Listening on 127.0.0.1:4444
1602513394779   webdriver::server       DEBUG   -> POST /session {"capabilities":{"javascriptEnabled":true,"nativeEvents":true,"acceptInsecureCerts":true}}
1602513394781   geckodriver::marionette DEBUG   Waiting 60s to connect to browser on 127.0.0.1:2828
1602513394783   geckodriver::marionette DEBUG   Connection to Marionette established on 127.0.0.1:2828.
1602513394795   webdriver::server       DEBUG   <- 200 OK {"value":{"sessionId":"1e86a24b-7248-47b3-a96e-776131d2edb0","capabilities":{"acceptInsecureCerts":false,"browserName":"firefox","browserVersion":"81.0.1","moz:accessibilityChecks":false,"moz:buildID":"20200930150533","moz:geckodriverVersion":"0.27.0","moz:headless":false,"moz:processID":63620,"moz:profile":"C:\\Users\\danpollak\\AppData\\Roaming\\Mozilla\\Firefox\\Profiles\\9e7wtvmu.test","moz:shutdownTimeout":60000,"moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"windows","platformVersion":"10.0","rotatable":false,"setWindowRect":true,"strictFileInteractability":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"unhandledPromptBehavior":"dismiss and notify"}}}
1602513394816   webdriver::server       DEBUG   -> GET /session/1e86a24b-7248-47b3-a96e-776131d2edb0/window/handles
1602513394819   webdriver::server       DEBUG   <- 500 Internal Server Error {"value":{"error":"unknown error","message":"TypeError: browser.browsingContext is null","stacktrace":"GeckoDriver.prototype.getIdForBrowser@chrome://marionette/content/driver.js:1477:15\nget@chrome://marionette/content/driver.js:273:28\nGeckoDriver.prototype.getWindowHandles@chrome://marionette/content/driver.js:1517:3\ndespatch@chrome://marionette/content/server.js:305:40\nexecute@chrome://marionette/content/server.js:275:16\nonPacket/<@chrome://marionette/content/server.js:248:20\nonPacket@chrome://marionette/content/server.js:249:9\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:501:20\n"}}
```

**Comment by whimboo** [comment]
Again, these are not trace level logs. Please read the link I pasted above correctly, and then provide the full log. Thanks.

**Comment by Danp2** [comment]
@whimboo Is this better?

```
1602544047741	geckodriver	INFO	Listening on 127.0.0.1:4444
1602544060052	webdriver::server	DEBUG	-> POST /session {"capabilities": {"alwaysMatch": {"browserName": "firefox", "acceptInsecureCerts":true, "moz:firefoxOptions": {"log": {"level": "trace"}}}}}
1602544060053	geckodriver::capabilities	DEBUG	Trying to read firefox version from ini files
1602544060053	geckodriver::capabilities	DEBUG	Found version 81.0.1
1602544060054	geckodriver::marionette	DEBUG	Waiting 60s to connect to browser on 127.0.0.1:2828
1602544060056	geckodriver::marionette	DEBUG	Connection to Marionette established on 127.0.0.1:2828.
1602544060068	webdriver::server	DEBUG	<- 200 OK {"value":{"sessionId":"933f61d7-0afb-42eb-865a-234fcf841866","capabilities":{"acceptInsecureCerts":true,"browserName":"firefox","browserVersion":"81.0.1","moz:accessibilityChecks":false,"moz:buildID":"20200930150533","moz:geckodriverVersion":"0.27.0","moz:headless":false,"moz:processID":63620,"moz:profile":"C:\\Users\\danpollak\\AppData\\Roaming\\Mozilla\\Firefox\\Profiles\\9e7wtvmu.test","moz:shutdownTimeout":60000,"moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"windows","platformVersion":"10.0","rotatable":false,"setWindowRect":true,"strictFileInteractability":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"unhandledPromptBehavior":"dismiss and notify"}}}
1602544060091	webdriver::server	DEBUG	-> GET /session/933f61d7-0afb-42eb-865a-234fcf841866/window/handles 
1602544060094	webdriver::server	DEBUG	<- 500 Internal Server Error {"value":{"error":"unknown error","message":"TypeError: browser.browsingContext is null","stacktrace":"GeckoDriver.prototype.getIdForBrowser@chrome://marionette/content/driver.js:1477:15\nget@chrome://marionette/content/driver.js:273:28\nGeckoDriver.prototype.getWindowHandles@chrome://marionette/content/driver.js:1517:3\ndespatch@chrome://marionette/content/server.js:305:40\nexecute@chrome://marionette/content/server.js:275:16\nonPacket/<@chrome://marionette/content/server.js:248:20\nonPacket@chrome://marionette/content/server.js:249:9\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:501:20\n"}}
```

Frankly, I find your linked reference confusing / inaccurate. It states --

> The log level is set in different ways. Either by using the --log <LEVEL> option, where LEVEL is one of the log levels from the list above, or by using the -v (for debug) or -vv (for trace) shorthands. 

This is what I tried in my first two attempts, which you state are not full trace logs. 

> The second way of setting the log level is through capabilities. geckodriver accepts a Mozilla-specific configuration object in moz:firefoxOptions.

For this most recent attempt, I used this method, which produced a few additional lines of output.

**Comment by whimboo** [comment]
That's good to know! Given that 82.0 will be released soon, we shouldn't worry about 81.0 anymore. I'm fairly sure that https://bugzilla.mozilla.org/show_bug.cgi?id=1493108 actually fixed that.
